File: android/src/main/kotlin/com/snowplowanalytics/snowplow_flutter_tracker/readers/events/ConsentGrantedReader.kt
package com.snowplowanalytics.snowplow_flutter_tracker.readers.events

import com.snowplowanalytics.snowplow.event.ConsentDocument
import com.snowplowanalytics.snowplow.event.ConsentGranted

class ConsentGrantedReader(val values: Map<String, Any>) {
    private val valuesDefault = values.withDefault { null }

    val expiry: String by values
    val documentId: String by values
    val version: String by values
    val name: String? by valuesDefault
    val documentDescription: String? by valuesDefault
    val consentDocuments: List<Map<String, Any>>? by valuesDefault
    private val processedConsentDocuments: List<ConsentDocument>? by lazy {
        consentDocuments?.let { it.map { ConsentDocumentReader(it).toConsentDocument() } }
    }

    fun toConsentGranted(): ConsentGranted {
        val consentGranted = ConsentGranted(expiry, documentId, version)
        name?.let { consentGranted.documentName(it) }
        documentDescription?.let { consentGranted.documentDescription(it) }
        processedConsentDocuments?.let { consentGranted.documents(it) }
        return consentGranted
    }
}

File: ios/Classes/readers/events/ConsentGrantedReader.swift
//
//  ConsentGrantedReader.swift
//  snowplow_flutter_tracker
//
//  Created by Matus Tomlein on 04/01/2022.
//

import Foundation
import SnowplowTracker

struct ConsentGrantedReader: Decodable {
    let expiry: String
    let documentId: String
    let version: String
    let name: String?
    let documentDescription: String?
    let consentDocuments: [ConsentDocumentReader]?
}

extension ConsentGrantedReader {
    func toConsentGranted() -> ConsentGranted {
        let event = ConsentGranted(expiry: expiry, documentId: documentId, version: version)
        if let name = self.name { event.name(name) }
        if let description = self.documentDescription { event.documentDescription(description) }
        if let documents = self.consentDocuments {
            let jsons = documents.map { $0.toConsentDocument().getPayload() }
            event.documents(jsons)
        }
        return event
    }
}
